### ISSUE TITLE
Cannot make projections of SPH data using non-region sources in yt-4.0

### ISSUE THREAD

**Issue by jzuhone** [yt-project/yt#2209]
### Bug report

**Bug summary**

The `sph_answer` test executes `PixelizedProjectionValuesTest` tests, which use a sphere as a selection object. This currently does not work for SPH data in yt-4.0. One gets error messages like this:

```
======================================================================
ERROR: PixelizedProjectionValues_snap_505_sphere_('c', (0_1, 'unitary'))_('gas', 'temperature')_2_('gas', 'density')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jzuhone/miniconda3/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 354, in __call__
    nv = self.run()
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 580, in run
    frb[self.field]
  File "/Users/jzuhone/Source/yt/yt/visualization/fixed_resolution.py", line 135, in __getitem__
    int(self.antialias))
  File "/Users/jzuhone/Source/yt/yt/geometry/coordinates/cartesian_coordinates.py", line 205, in pixelize
    antialias, dimension, periodic)
  File "/Users/jzuhone/Source/yt/yt/geometry/coordinates/cartesian_coordinates.py", line 300, in _ortho_pixelize
    le = data_source.data_source.left_edge.in_units('code_length')
AttributeError: 'YTSphere' object has no attribute 'left_edge'

======================================================================
ERROR: FieldValues_snap_505_sphere_('c', (0_1, 'unitary'))_('gas', 'temperature')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jzuhone/miniconda3/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 354, in __call__
    nv = self.run()
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 471, in run
    field, weight=weight_field)
  File "/Users/jzuhone/Source/yt/yt/data_objects/derived_quantities.py", line 147, in __call__
    rv = super(WeightedAverageQuantity, self).__call__(fields, weight)
  File "/Users/jzuhone/Source/yt/yt/data_objects/derived_quantities.py", line 75, in __call__
    sto.result = self.process_chunk(ds, *args, **kwargs)
  File "/Users/jzuhone/Source/yt/yt/data_objects/derived_quantities.py", line 153, in process_chunk
    for field in fields]
  File "/Users/jzuhone/Source/yt/yt/data_objects/derived_quantities.py", line 153, in <listcomp>
    for field in fields]
  File "/Users/jzuhone/Source/yt/yt/data_objects/data_containers.py", line 257, in __getitem__
    self.get_data(f)
  File "/Users/jzuhone/Source/yt/yt/data_objects/data_containers.py", line 1417, in get_data
    self._generate_fields(fields_to_generate)
  File "/Users/jzuhone/Source/yt/yt/data_objects/data_containers.py", line 1437, in _generate_fields
    fd = self._generate_field(field)
  File "/Users/jzuhone/Source/yt/yt/data_objects/data_containers.py", line 294, in _generate_field
    tr = self._generate_fluid_field(field)
  File "/Users/jzuhone/Source/yt/yt/data_objects/data_containers.py", line 314, in _generate_fluid_field
    rv = finfo(gen_obj)
  File "/Users/jzuhone/Source/yt/yt/fields/derived_field.py", line 275, in __call__
    dd = self._function(self, data)
  File "/Users/jzuhone/Source/yt/yt/fields/geometric_fields.py", line 112, in _ones
    arr = np.ones(data.ires.shape, dtype="float64")
  File "/Users/jzuhone/Source/yt/yt/data_objects/data_containers.py", line 1602, in ires
    return self._current_chunk.ires
  File "/Users/jzuhone/Source/yt/yt/geometry/geometry_handler.py", line 269, in cached_func
    tr = self._accumulate_values(n[1:])
  File "/Users/jzuhone/Source/yt/yt/geometry/geometry_handler.py", line 297, in _accumulate_values
    arrs.append(f(self.dobj))
  File "/Users/jzuhone/Source/yt/yt/data_objects/particle_container.py", line 30, in _func_non_indexed
    raise YTNonIndexedDataContainer(self)
yt.utilities.exceptions.YTNonIndexedDataContainer: The data container (<class 'yt.data_objects.particle_container.ParticleContainer'>) is an unindexed type.  Operations such as ires, icoords, fcoords and fwidth will not work on it.

======================================================================
ERROR: PixelizedProjectionValues_snap_505_sphere_('c', (0_1, 'unitary'))_('gas', 'velocity_magnitude')_0_None
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jzuhone/miniconda3/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 354, in __call__
    nv = self.run()
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 580, in run
    frb[self.field]
  File "/Users/jzuhone/Source/yt/yt/visualization/fixed_resolution.py", line 135, in __getitem__
    int(self.antialias))
  File "/Users/jzuhone/Source/yt/yt/geometry/coordinates/cartesian_coordinates.py", line 205, in pixelize
    antialias, dimension, periodic)
  File "/Users/jzuhone/Source/yt/yt/geometry/coordinates/cartesian_coordinates.py", line 300, in _ortho_pixelize
    le = data_source.data_source.left_edge.in_units('code_length')
AttributeError: 'YTSphere' object has no attribute 'left_edge'

======================================================================
ERROR: PixelizedProjectionValues_snap_505_sphere_('c', (0_1, 'unitary'))_('gas', 'velocity_magnitude')_1_None
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jzuhone/miniconda3/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 354, in __call__
    nv = self.run()
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 580, in run
    frb[self.field]
  File "/Users/jzuhone/Source/yt/yt/visualization/fixed_resolution.py", line 135, in __getitem__
    int(self.antialias))
  File "/Users/jzuhone/Source/yt/yt/geometry/coordinates/cartesian_coordinates.py", line 205, in pixelize
    antialias, dimension, periodic)
  File "/Users/jzuhone/Source/yt/yt/geometry/coordinates/cartesian_coordinates.py", line 300, in _ortho_pixelize
    le = data_source.data_source.left_edge.in_units('code_length')
AttributeError: 'YTSphere' object has no attribute 'left_edge'

======================================================================
ERROR: PixelizedProjectionValues_snap_505_sphere_('c', (0_1, 'unitary'))_('gas', 'velocity_magnitude')_2_None
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jzuhone/miniconda3/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 354, in __call__
    nv = self.run()
  File "/Users/jzuhone/Source/yt/yt/utilities/answer_testing/framework.py", line 580, in run
    frb[self.field]
  File "/Users/jzuhone/Source/yt/yt/visualization/fixed_resolution.py", line 135, in __getitem__
    int(self.antialias))
  File "/Users/jzuhone/Source/yt/yt/geometry/coordinates/cartesian_coordinates.py", line 205, in pixelize
    antialias, dimension, periodic)
  File "/Users/jzuhone/Source/yt/yt/geometry/coordinates/cartesian_coordinates.py", line 300, in _ortho_pixelize
    le = data_source.data_source.left_edge.in_units('code_length')
AttributeError: 'YTSphere' object has no attribute 'left_edge'
```

**Version Information**
  * Operating System: macOS
  * Python Version: 3.7
  * yt version: 4.0-dev

**Comment by ngoldbaum** [comment]
This is due to this bit in the SPH projection pixelizer:

https://github.com/yt-project/yt/blob/yt-4.0/yt/geometry/coordinates/cartesian_coordinates.py#L298-L312

As you can see if the projection's data source is anything besides a region this won't work.

Looking at the code it's not clear to me why this indirection is necessary - shouldn't the projection be doing I/O using the underlying data source? Probably worth investigating what happens when you remove the creation of `proj_reg`.

**Comment by ngoldbaum** [comment]
Oh wait, I see, so here's what's happening. Imagine you're doing a projection of the entire domain, but are zoomed in on something interesting in the center of the domain. Without this hack then you'd read in *all* of the particle data across the whole domain, with this hack you only read in the particles that contribute to the image.

So there are two things we could do. The first, easy thing would be to check if we're handed anything besides a region and don't do this optimization in that case.

Another option would be to go through the data objects and add a new method that calculates the bounds of the smallest axis-aligned bounding box for each object. Then here we could just call that function.

**Comment by jzuhone** [comment]
> Another option would be to go through the data objects and add a new method that calculates the bounds of the smallest axis-aligned bounding box for each object. Then here we could just call that function.

I tried to do something like this in pyXSIM for general objects, but I wasn't smart enough to figure it out or it's too hard to come up with something that can handle all of the general cases (tilted disks, for example).

**Comment by ngoldbaum** [comment]
I'm pretty sure a formula can be derived for all of the geometric objects and are available in the game or 3d graphics programming literature. For example, here's one for inclined cylinders: http://www.iquilezles.org/www/articles/diskbbox/diskbbox.htm

For cut_regions we could delegate to the underlying geometric data objects.

**Comment by qobilidop** [comment]
Closed by #2215.
